<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Tutorial • Gretool</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Gretool</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/Tutorial.html">Gretool Tutorial</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="nav-link" href="../../"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/thereallda/Gretool"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Tutorial</h1>
            
      

      <div class="d-none name"><code>Tutorial.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>Gretool</code> package is implemented in R and publicly
available at <a href="https://github.com/thereallda/Gretool" class="external-link uri">https://github.com/thereallda/Gretool</a>.</p>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<p>For this tutorial, we will demonstrate the <code>Gretool</code>
workflow by using a glyco-capped RNA-seq data from human.</p>
<p>We start by reading in the data, including count matrix and metadata.
The metadata requires at lease two columns:</p>
<p>1. The <em>condition</em> column specifies the biological group of
each sample, such as <strong>A.Input</strong>, indicating that the
sample is an input sample from biological condition A.</p>
<p>2. The <em>enrich</em> column denotes the enrichment group of each
sample, such as <strong>Input</strong>, representing the input
library.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://thereallda.github.io/Gretool/">Gretool</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># read in metadata and counts matrix</span></span>
<span><span class="va">counts_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"data/Counts.csv"</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"data/metadata.csv"</span>, comment.char <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span></span>
<span><span class="co">#&gt;         id sample.id condition enrich replicate</span></span>
<span><span class="co">#&gt; 1 H.input1  H.input1   H.Input  Input         1</span></span>
<span><span class="co">#&gt; 2 H.input2  H.input2   H.Input  Input         2</span></span>
<span><span class="co">#&gt; 3 H.input3  H.input3   H.Input  Input         3</span></span>
<span><span class="co">#&gt; 4 H.input4  H.input4   H.Input  Input         4</span></span>
<span><span class="co">#&gt; 5    H.ip1     H.ip1  H.Enrich Enrich         1</span></span>
<span><span class="co">#&gt; 6    H.ip2     H.ip2  H.Enrich Enrich         2</span></span>
<span><span class="co"># rownames of metadata should be consistent with the colnames of counts_mat</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">meta</span><span class="op">$</span><span class="va">id</span></span></code></pre></div>
<p>Also, other information, such as batch group can be included in the
metadata.</p>
<p>Next, we use the count matrix and metadata to create a
<code>Gretool</code> object.</p>
<p>The <code>Gretool</code> object serves as a container that contains
both data (raw and normalized counts data) and analysis (e.g.,
normalization performance evaluation score and enrichment results).</p>
<p>When constructing <code>Gretool</code> object, we can provide
following arguments:</p>
<ul>
<li><p>The id of input (<code>input.id = "Input"</code>) and enrichment
(<code>enrich.id = "Enrich"</code>), same as the enrich column in the
metadata (<strong>required</strong>).</p></li>
<li><p>Prefix of spike-in genes (<code>spike.in.prefix = "^FB"</code>)
(optional);</p></li>
<li><p>The id of synthetic spike-in
(<code>synthetic.id = c("Syn1", "Syn2")</code>), same with row names in
<code>counts_mat</code> (optional);</p></li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create object</span></span>
<span><span class="va">Gretool</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createGretool.html">createGretool</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">counts_mat</span>,</span>
<span>                         col.data <span class="op">=</span> <span class="va">meta</span>,</span>
<span>                         input.id <span class="op">=</span> <span class="st">"Input"</span>,</span>
<span>                         enrich.id <span class="op">=</span> <span class="st">"Enrich"</span></span>
<span>                         <span class="op">)</span></span>
<span><span class="va">Gretool</span></span>
<span><span class="co">#&gt; class: Gretool </span></span>
<span><span class="co">#&gt; dim: 62700 16 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): ''</span></span>
<span><span class="co">#&gt; rownames(62700): ENSG00000279928 ENSG00000228037 ... ENSG00000210195</span></span>
<span><span class="co">#&gt;   ENSG00000210196</span></span>
<span><span class="co">#&gt; rowData names(3): GeneID SpikeIn Synthetic</span></span>
<span><span class="co">#&gt; colnames(16): H.input1 H.input2 ... T.ip3 T.ip4</span></span>
<span><span class="co">#&gt; colData names(7): id id.1 ... replicate batch</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p><code>Gretool</code> starts with quality control step to filter
lowly-expressed genes. This step is performed with
<code>FilterLowExprGene</code> by keeping genes with at least
<code>min.count</code> in at least n samples. n is determined by the
smallest group sample size specifying in <code>group</code> .</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Gretool</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FilterLowExprGene.html">FilterLowExprGene</a></span><span class="op">(</span><span class="va">Gretool</span>, group <span class="op">=</span> <span class="va">Gretool</span><span class="op">$</span><span class="va">condition</span>, min.count <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">Gretool</span></span>
<span><span class="co">#&gt; class: Gretool </span></span>
<span><span class="co">#&gt; dim: 15832 16 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): ''</span></span>
<span><span class="co">#&gt; rownames(15832): ENSG00000142611 ENSG00000157911 ... ENSG00000210195</span></span>
<span><span class="co">#&gt;   ENSG00000210196</span></span>
<span><span class="co">#&gt; rowData names(3): GeneID SpikeIn Synthetic</span></span>
<span><span class="co">#&gt; colnames(16): H.input1 H.input2 ... T.ip3 T.ip4</span></span>
<span><span class="co">#&gt; colData names(7): id id.1 ... replicate batch</span></span></code></pre></div>
<div class="section level3">
<h3 id="normalization-and-assessment">Normalization and assessment<a class="anchor" aria-label="anchor" href="#normalization-and-assessment"></a>
</h3>
<p>The main <code>Gretool</code> function can be used to perform gene
selection, normalization, and evaluation. This function return selected
gene set in <code>rowData</code>, normalized count matrix in
<code>counts</code> slot (when <code>return.norm=TRUE</code>),
normalization factors (<code>norm_factors</code> slot), evaluation
metrics and scores in <code>norm_metrics</code> and
<code>norm_scores</code> slots, respectively. Detailed of these steps
are described bellow.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Gretool</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Gretool.html">Gretool</a></span><span class="op">(</span><span class="va">Gretool</span>, </span>
<span>               scaling.method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="st">"TMM"</span>, <span class="st">"DESeq"</span><span class="op">)</span>,</span>
<span>               ruv.norm <span class="op">=</span> <span class="cn">TRUE</span>, ruv.k <span class="op">=</span> <span class="fl">3</span>,</span>
<span>               eval.pam.k <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">3</span>, eval.pc.n <span class="op">=</span> <span class="fl">3</span>, </span>
<span>               return.norm <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>               <span class="op">)</span></span>
<span><span class="co">#&gt; Gene set selection for normalization and assessment...</span></span>
<span><span class="co">#&gt; - The number of negative control genes for normalization: 1000 </span></span>
<span><span class="co">#&gt; - Estimate dispersion &amp; Fit GLM... </span></span>
<span><span class="co">#&gt; - Testing differential genes... </span></span>
<span><span class="co">#&gt; - The number of positive evaluation genes: 500 </span></span>
<span><span class="co">#&gt; - Estimate dispersion &amp; Fit GLM... </span></span>
<span><span class="co">#&gt; - Testing differential genes... </span></span>
<span><span class="co">#&gt; - The number of negative evaluation genes: 500 </span></span>
<span><span class="co">#&gt; - Estimate dispersion &amp; Fit GLM... </span></span>
<span><span class="co">#&gt; - Testing differential genes... </span></span>
<span><span class="co">#&gt; Apply normalization...</span></span>
<span><span class="co">#&gt; - Scaling... </span></span>
<span><span class="co">#&gt; - Regression-based normalization... </span></span>
<span><span class="co">#&gt; Perform assessment...</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="effect-of-normalization">Effect of normalization<a class="anchor" aria-label="anchor" href="#effect-of-normalization"></a>
</h2>
<p>Here, we use the top-ranked procedure DESeq_RUVs_k3 for downstream
analysis.</p>
<p>We perform PCA based on the count matrix from sample of interest
before and after the normalization, for demonstrating the effect of
normalization.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get scores for normalization</span></span>
<span><span class="va">norm_scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getScore.html">getScore</a></span><span class="op">(</span><span class="va">Gretool</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">norm_scores</span><span class="op">)</span></span>
<span><span class="co">#&gt;               BIO_SIM EN_SIM BATCH_SIM PAM_SIM RLE_MED RLE_IQR WV_COR UV_COR</span></span>
<span><span class="co">#&gt; DESeq_RUVs_k3      28     29         1      29      30      14     26     28</span></span>
<span><span class="co">#&gt; TMM_RUVs_k3        29     28         1      28      29      15     27     27</span></span>
<span><span class="co">#&gt; Raw_RUVs_k3        30     30         1      30       7      30     22     15</span></span>
<span><span class="co">#&gt; DESeq_RUVs_k2      27     22         1      27      28       8     23     16</span></span>
<span><span class="co">#&gt; TMM_RUVs_k2        26     19         1      26      27      13     20     11</span></span>
<span><span class="co">#&gt; TMM_RUVs_k1        23     14         1      20      24      25      9     20</span></span>
<span><span class="co">#&gt;                  SCORE</span></span>
<span><span class="co">#&gt; DESeq_RUVs_k3 26.28571</span></span>
<span><span class="co">#&gt; TMM_RUVs_k3   26.14286</span></span>
<span><span class="co">#&gt; Raw_RUVs_k3   23.42857</span></span>
<span><span class="co">#&gt; DESeq_RUVs_k2 21.57143</span></span>
<span><span class="co">#&gt; TMM_RUVs_k2   20.28571</span></span>
<span><span class="co">#&gt; TMM_RUVs_k1   19.28571</span></span>
<span></span>
<span><span class="co"># select normalization</span></span>
<span><span class="va">norm_method</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">norm_scores</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get normalized counts</span></span>
<span><span class="va">norm_data</span> <span class="op">&lt;-</span> <span class="fu">Gretool</span><span class="fu">::</span><span class="fu"><a href="../reference/Counts.html">Counts</a></span><span class="op">(</span><span class="va">Gretool</span>, slot <span class="op">=</span> <span class="st">"sample"</span>, method <span class="op">=</span> <span class="va">norm_method</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get normalization factors</span></span>
<span><span class="va">norm_factors</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getFactor.html">getFactor</a></span><span class="op">(</span><span class="va">Gretool</span>, slot <span class="op">=</span> <span class="st">"sample"</span>, method <span class="op">=</span> <span class="va">norm_method</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create sample name, e.g., A.Input.1</span></span>
<span><span class="va">samples_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">Gretool</span><span class="op">$</span><span class="va">condition</span>, <span class="va">Gretool</span><span class="op">$</span><span class="va">replicate</span>, sep<span class="op">=</span><span class="st">"."</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># PCA for raw count</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PCAplot.html">PCAplot</a></span><span class="op">(</span><span class="fu">Gretool</span><span class="fu">::</span><span class="fu"><a href="../reference/Counts.html">Counts</a></span><span class="op">(</span><span class="va">Gretool</span>, slot<span class="op">=</span><span class="st">"sample"</span>, <span class="st">"Raw"</span><span class="op">)</span>, </span>
<span>            color <span class="op">=</span> <span class="va">Gretool</span><span class="op">$</span><span class="va">condition</span>,</span>
<span>            shape <span class="op">=</span> <span class="va">Gretool</span><span class="op">$</span><span class="va">enrich</span>,</span>
<span>            label <span class="op">=</span> <span class="va">samples_name</span>, </span>
<span>            vst.norm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Raw"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: `aes_string()` was deprecated in ggplot2 3.0.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use tidy evaluation idioms with `aes()`.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> See also `vignette("ggplot2-in-packages")` for more information.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">Gretool</span> package.</span></span>
<span><span class="co">#&gt;   Please report the issue to the authors.</span></span>
<span><span class="co">#&gt; This warning is displayed once every 8 hours.</span></span>
<span><span class="co">#&gt; Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span>
<span><span class="co">#&gt; generated.</span></span>
<span></span>
<span><span class="co"># PCA for normalized count</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PCAplot.html">PCAplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">norm_data</span><span class="op">)</span>, </span>
<span>            color <span class="op">=</span> <span class="va">Gretool</span><span class="op">$</span><span class="va">condition</span>,</span>
<span>            shape <span class="op">=</span> <span class="va">Gretool</span><span class="op">$</span><span class="va">enrich</span>,</span>
<span>            label <span class="op">=</span> <span class="va">samples_name</span>, </span>
<span>            vst.norm <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"After normalization"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># combine two plots</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="Tutorial_files/figure-html/unnamed-chunk-6-1.png" class="r-plt" alt="" width="1152"></p>
</div>
<div class="section level2">
<h2 id="find-enrichment">Find enrichment<a class="anchor" aria-label="anchor" href="#find-enrichment"></a>
</h2>
<p><code>Gretool</code> package can help you find enrichment genes from
each biological groups via differential expression. It can identify
genes that significantly increased in enrichment samples compared to
input samples.</p>
<p><code>FindEnrichment</code> automates this processes for all groups
provided in <code>condition</code> column.</p>
<p>By default, enriched genes are defined as fold change of normalized
transcript counts ≥ 2 (<code>logfc.cutoff = 1</code>), FDR &lt; 0.05
(<code>p.cutoff = 0.05</code>) in enrichment samples compared to those
in input samples.</p>
<p>Use <code>getEnrichment</code> to retrieve a list of enrichment
result tables.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># find all enriched genes</span></span>
<span><span class="va">Gretool</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindEnrichment.html">FindEnrichment</a></span><span class="op">(</span><span class="va">Gretool</span>, </span>
<span>                          slot<span class="op">=</span><span class="st">"sample"</span>, </span>
<span>                          norm.method <span class="op">=</span> <span class="va">norm_method</span>, </span>
<span>                          logfc.cutoff <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                          p.cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="co">#&gt; - Estimate dispersion &amp; Fit GLM... </span></span>
<span><span class="co">#&gt; - Testing differential genes...</span></span>
<span></span>
<span><span class="co"># get filtered enrichment results</span></span>
<span><span class="va">res.sig.ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getEnrichment.html">getEnrichment</a></span><span class="op">(</span><span class="va">Gretool</span>, slot<span class="op">=</span><span class="st">"sample"</span>, filter<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">res.sig.ls</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'H'</span>,<span class="st">'T'</span><span class="op">)</span></span>
<span><span class="co"># count number of enrichment in each group</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">res.sig.ls</span>, <span class="va">nrow</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;    H    T </span></span>
<span><span class="co">#&gt; 1924 2023</span></span></code></pre></div>
<p>Each enrichment table is a <code>data.frame</code> with a list of
genes as rows, and associated information as columns (GeneID, logFC,
p-values, etc.). The following columns are present in the table:</p>
<ul>
<li>
<code>GeneID</code>: ID of genes.</li>
<li>
<code>logFC</code>: log2 fold-change between enrichment and input
samples. Positive values indicate that the gene is more highly enriched
in the enrichment group.</li>
<li>
<code>logCPM</code>: log2 CPM (counts per million) of the average
expression of all samples.</li>
<li>
<code>LR</code>: Likelihood ratio of the likelihood ratio test.</li>
<li>
<code>PValue</code>: p-value from the likelihood ratio test.</li>
<li>
<code>FDR</code>: False discovery rate of the p-value, default “BH”
method is applied.</li>
</ul>
</div>
<div class="section level2">
<h2 id="differential-modification-analysis">Differential modification analysis<a class="anchor" aria-label="anchor" href="#differential-modification-analysis"></a>
</h2>
<p>You can perform differential modification analysis with
<code>DiffModification</code></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create biological column in metadata</span></span>
<span><span class="va">meta</span><span class="op">$</span><span class="va">bio</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">'\\..*'</span>,<span class="st">''</span>,<span class="va">meta</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span></span>
<span><span class="va">meta</span></span>
<span><span class="co">#&gt;                id sample.id condition enrich replicate bio</span></span>
<span><span class="co">#&gt; H.input1 H.input1  H.input1   H.Input  Input         1   H</span></span>
<span><span class="co">#&gt; H.input2 H.input2  H.input2   H.Input  Input         2   H</span></span>
<span><span class="co">#&gt; H.input3 H.input3  H.input3   H.Input  Input         3   H</span></span>
<span><span class="co">#&gt; H.input4 H.input4  H.input4   H.Input  Input         4   H</span></span>
<span><span class="co">#&gt; H.ip1       H.ip1     H.ip1  H.Enrich Enrich         1   H</span></span>
<span><span class="co">#&gt; H.ip2       H.ip2     H.ip2  H.Enrich Enrich         2   H</span></span>
<span><span class="co">#&gt; H.ip3       H.ip3     H.ip3  H.Enrich Enrich         3   H</span></span>
<span><span class="co">#&gt; H.ip4       H.ip4     H.ip4  H.Enrich Enrich         4   H</span></span>
<span><span class="co">#&gt; T.input1 T.input1  T.input1   T.Input  Input         1   T</span></span>
<span><span class="co">#&gt; T.input2 T.input2  T.input2   T.Input  Input         2   T</span></span>
<span><span class="co">#&gt; T.input3 T.input3  T.input3   T.Input  Input         3   T</span></span>
<span><span class="co">#&gt; T.input4 T.input4  T.input4   T.Input  Input         4   T</span></span>
<span><span class="co">#&gt; T.ip1       T.ip1     T.ip1  T.Enrich Enrich         1   T</span></span>
<span><span class="co">#&gt; T.ip2       T.ip2     T.ip2  T.Enrich Enrich         2   T</span></span>
<span><span class="co">#&gt; T.ip3       T.ip3     T.ip3  T.Enrich Enrich         3   T</span></span>
<span><span class="co">#&gt; T.ip4       T.ip4     T.ip4  T.Enrich Enrich         4   T</span></span>
<span></span>
<span><span class="co"># testing </span></span>
<span><span class="va">diff_mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DiffModification.html">DiffModification</a></span><span class="op">(</span></span>
<span>  <span class="va">Gretool</span>,</span>
<span>  norm.method <span class="op">=</span> <span class="va">norm_method</span>,</span>
<span>  col.data <span class="op">=</span> <span class="va">meta</span>,</span>
<span>  contrast.df <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>Group1<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'H'</span><span class="op">)</span>, </span>
<span>                           Group2<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'T'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  assay.colname <span class="op">=</span> <span class="st">'enrich'</span>,</span>
<span>  biology.colname <span class="op">=</span> <span class="st">'bio'</span>,</span>
<span>  input.id <span class="op">=</span> <span class="st">'Input'</span>,</span>
<span>  enrich.id <span class="op">=</span> <span class="st">'Enrich'</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">1e-4</span>,</span>
<span>  shrink_lambda <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; - Estimate dispersion &amp; Fit GLM... </span></span>
<span><span class="co">#&gt; - Testing differential genes...</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">diff_mod</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;            GeneID  H.logFC         H.FDR  T.logFC        T.FDR        LR</span></span>
<span><span class="co">#&gt; 1 ENSG00000210196 3.369089 5.975863e-167 2.400430 3.665593e-92 1.8006983</span></span>
<span><span class="co">#&gt; 4 ENSG00000212402 8.407736  4.578186e-89 5.930855 3.613089e-50 0.8825648</span></span>
<span><span class="co">#&gt; 5 ENSG00000263934 3.226858  1.271129e-87 3.001821 3.209291e-81 2.4853247</span></span>
<span><span class="co">#&gt; 6 ENSG00000136159 2.221450  5.022980e-81 2.249885 2.194611e-89 0.6462647</span></span>
<span><span class="co">#&gt; 7 ENSG00000198786 2.528707  2.740577e-79 2.228075 1.809955e-62 0.1451544</span></span>
<span><span class="co">#&gt; 8 ENSG00000173214 1.661057  9.826775e-73 1.726249 5.285824e-65 0.9099521</span></span>
<span><span class="co">#&gt;      PValue   mod.FDR  logFC.diff mod.type</span></span>
<span><span class="co">#&gt; 1 0.1796281 0.4170512  0.96865947       NS</span></span>
<span><span class="co">#&gt; 4 0.3475002 0.5457629  2.47688160       NS</span></span>
<span><span class="co">#&gt; 5 0.1149126 0.3478914  0.22503734       NS</span></span>
<span><span class="co">#&gt; 6 0.4214513 0.5947956 -0.02843545       NS</span></span>
<span><span class="co">#&gt; 7 0.7032096 0.7958550  0.30063236       NS</span></span>
<span><span class="co">#&gt; 8 0.3401271 0.5406134 -0.06519206       NS</span></span></code></pre></div>
<p>By default, <code>DiffModification</code> output all differential
modification results for all genes. You can subset the table based on
the enriched genes from <code>FindEnrichment</code></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enrich_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">res.sig.ls</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">$</span><span class="va">GeneID</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">diff_mod_enrich</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">diff_mod</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">GeneID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">enrich_ids</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">diff_mod_enrich</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 2599   10</span></span></code></pre></div>
<p>You can visualize the differential results with scatter plot</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">typedf1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">diff_mod_enrich</span><span class="op">$</span><span class="va">mod.type</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">Var1</span>,<span class="st">' (n='</span>,<span class="va">Freq</span>,<span class="st">')'</span><span class="op">)</span>,</span>
<span>         prop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fl">100</span><span class="op">*</span><span class="va">Freq</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">Freq</span><span class="op">)</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">diff_mod_enrich</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">T.logFC</span>,<span class="va">H.logFC</span>, color<span class="op">=</span><span class="va">mod.type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>size <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">mod.FDR</span><span class="op">)</span><span class="op">)</span>, alpha<span class="op">=</span><span class="fl">0.8</span>, shape<span class="op">=</span><span class="fl">16</span>, stroke<span class="op">=</span><span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">'right'</span>,</span>
<span>        axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">'black'</span><span class="op">)</span>,</span>
<span>        plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#A83C1E"</span>,<span class="st">"#4F99B4"</span>,<span class="st">"grey80"</span><span class="op">)</span>,</span>
<span>                     labels<span class="op">=</span><span class="va">typedf1</span><span class="op">$</span><span class="va">label</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">''</span>, </span>
<span>       title<span class="op">=</span><span class="st">'H vs T'</span><span class="op">)</span></span></code></pre></div>
<p><img src="Tutorial_files/figure-html/unnamed-chunk-10-1.png" class="r-plt" alt="" width="700"></p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.2 (2025-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] patchwork_1.3.2 tidyr_1.3.2     ggplot2_4.0.1   dplyr_1.1.4    </span></span>
<span><span class="co">#&gt; [5] Gretool_1.0.0  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] pbapply_1.7-4               rlang_1.1.6                </span></span>
<span><span class="co">#&gt;   [3] magrittr_2.0.4              otel_0.2.0                 </span></span>
<span><span class="co">#&gt;   [5] matrixStats_1.5.0           compiler_4.5.2             </span></span>
<span><span class="co">#&gt;   [7] flexmix_2.3-20              systemfonts_1.3.1          </span></span>
<span><span class="co">#&gt;   [9] vctrs_0.6.5                 stringr_1.6.0              </span></span>
<span><span class="co">#&gt;  [11] pkgconfig_2.0.3             fastmap_1.2.0              </span></span>
<span><span class="co">#&gt;  [13] backports_1.5.0             XVector_0.50.0             </span></span>
<span><span class="co">#&gt;  [15] labeling_0.4.3              rmarkdown_2.30             </span></span>
<span><span class="co">#&gt;  [17] ragg_1.5.0                  purrr_1.2.0                </span></span>
<span><span class="co">#&gt;  [19] xfun_0.55                   modeltools_0.2-24          </span></span>
<span><span class="co">#&gt;  [21] cachem_1.1.0                jsonlite_2.0.0             </span></span>
<span><span class="co">#&gt;  [23] EnvStats_3.1.0              DelayedArray_0.36.0        </span></span>
<span><span class="co">#&gt;  [25] fpc_2.2-13                  BiocParallel_1.44.0        </span></span>
<span><span class="co">#&gt;  [27] broom_1.0.11                parallel_4.5.2             </span></span>
<span><span class="co">#&gt;  [29] prabclus_2.3-4              cluster_2.1.8.1            </span></span>
<span><span class="co">#&gt;  [31] R6_2.6.1                    bslib_0.9.0                </span></span>
<span><span class="co">#&gt;  [33] stringi_1.8.7               RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;  [35] limma_3.66.0                car_3.1-3                  </span></span>
<span><span class="co">#&gt;  [37] GenomicRanges_1.62.1        jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt;  [39] diptest_0.77-2              Rcpp_1.1.0                 </span></span>
<span><span class="co">#&gt;  [41] Seqinfo_1.0.0               SummarizedExperiment_1.40.0</span></span>
<span><span class="co">#&gt;  [43] knitr_1.51                  IRanges_2.44.0             </span></span>
<span><span class="co">#&gt;  [45] Matrix_1.7-4                nnet_7.3-20                </span></span>
<span><span class="co">#&gt;  [47] tidyselect_1.2.1            abind_1.4-8                </span></span>
<span><span class="co">#&gt;  [49] yaml_2.3.12                 codetools_0.2-20           </span></span>
<span><span class="co">#&gt;  [51] lattice_0.22-7              tibble_3.3.0               </span></span>
<span><span class="co">#&gt;  [53] Biobase_2.70.0              withr_3.0.2                </span></span>
<span><span class="co">#&gt;  [55] S7_0.2.1                    evaluate_1.0.5             </span></span>
<span><span class="co">#&gt;  [57] desc_1.4.3                  mclust_6.1.2               </span></span>
<span><span class="co">#&gt;  [59] kernlab_0.9-33              pillar_1.11.1              </span></span>
<span><span class="co">#&gt;  [61] ggpubr_0.6.2                MatrixGenerics_1.22.0      </span></span>
<span><span class="co">#&gt;  [63] carData_3.0-5               stats4_4.5.2               </span></span>
<span><span class="co">#&gt;  [65] plotly_4.11.0               generics_0.1.4             </span></span>
<span><span class="co">#&gt;  [67] S4Vectors_0.48.0            scales_1.4.0               </span></span>
<span><span class="co">#&gt;  [69] class_7.3-23                glue_1.8.0                 </span></span>
<span><span class="co">#&gt;  [71] lazyeval_0.2.2              tools_4.5.2                </span></span>
<span><span class="co">#&gt;  [73] robustbase_0.99-6           data.table_1.18.0          </span></span>
<span><span class="co">#&gt;  [75] locfit_1.5-9.12             ggsignif_0.6.4             </span></span>
<span><span class="co">#&gt;  [77] fs_1.6.6                    grid_4.5.2                 </span></span>
<span><span class="co">#&gt;  [79] edgeR_4.8.2                 Formula_1.2-5              </span></span>
<span><span class="co">#&gt;  [81] cli_3.6.5                   textshaping_1.0.4          </span></span>
<span><span class="co">#&gt;  [83] S4Arrays_1.10.1             viridisLite_0.4.2          </span></span>
<span><span class="co">#&gt;  [85] paintingr_0.1.0             gtable_0.3.6               </span></span>
<span><span class="co">#&gt;  [87] DEoptimR_1.1-4              rstatix_0.7.3              </span></span>
<span><span class="co">#&gt;  [89] DESeq2_1.50.2               sass_0.4.10                </span></span>
<span><span class="co">#&gt;  [91] digest_0.6.39               BiocGenerics_0.56.0        </span></span>
<span><span class="co">#&gt;  [93] SparseArray_1.10.8          ggrepel_0.9.6              </span></span>
<span><span class="co">#&gt;  [95] htmlwidgets_1.6.4           farver_2.1.2               </span></span>
<span><span class="co">#&gt;  [97] htmltools_0.5.9             pkgdown_2.2.0              </span></span>
<span><span class="co">#&gt;  [99] lifecycle_1.0.4             httr_1.4.7                 </span></span>
<span><span class="co">#&gt; [101] statmod_1.5.1               MASS_7.3-65</span></span></code></pre></div>
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Dean Li.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
